src/main/java/core/GraphReducer.java

```diff
@@ -66,12 +67,54 @@ public static void determineParents(HashMap<Integer, Node> nodeMap) {
             // than the number of nodes after previous reduction.
             if ((previousMapSize - currentMapSize) <= minDelta) {
                 levelMaps.set(i - 1, levelMap);
+                int maxDepth = 20;
+                for (int j = i; maxDepth < 1001; j++) {
```

Comment:
What is the 1001 for?

If this is just a random upper boundary, use a while loop instead.